File: csharp/linked-list/LinkedList.cs
public class Deque<T>
{
    private Node _firstNode;
    private Node _lastNode;

    public void Push(T value)
    {
        var oldLastNode = _lastNode;

        _lastNode = new Node(value) {Previous = oldLastNode};

        if (oldLastNode != null)
        {
            oldLastNode.Next = _lastNode;
        }

        _firstNode ??= _lastNode;
    }

    public T Pop()
    {
        var value = _lastNode.Value;
        _lastNode = _lastNode.Previous;
        return value;
    }

    public void Unshift(T value)
    {
        var oldFirstNode = _firstNode;

        _firstNode = new Node(value) {Next = oldFirstNode};

        if (oldFirstNode != null)
        {
            oldFirstNode.Previous = _firstNode;
        }

        _lastNode ??= _firstNode;
    }

    public T Shift()
    {
        var value = _firstNode.Value;
        _firstNode = _firstNode.Next;
        return value;
    }

    private class Node
    {
        public Node(T value)
        {
            Value = value;
        }

        public T Value { get; private set; }

        public Node Next { get; set; }

        public Node Previous { get; set; }
    }
}

File: fsharp/linked-list/LinkedList.fs
module LinkedList

type Node<'a> =
    { Value: 'a
      mutable Next: Node<'a> option
      mutable Previous: Node<'a> option }

type LinkedList<'a> =
    { mutable FirstNode: Node<'a> option
      mutable LastNode: Node<'a> option }

let mkLinkedList () = { FirstNode = None; LastNode = None }

let pop linkedList =
    let lastNode = Option.get linkedList.LastNode
    linkedList.LastNode <- lastNode.Previous
    lastNode.Value

let shift linkedList =
    let firstNode = Option.get linkedList.FirstNode
    linkedList.FirstNode <- firstNode.Next
    firstNode.Value

let push newValue linkedList =
    let lastNode = linkedList.LastNode

    linkedList.LastNode <-
        Some
            { Value = newValue
              Previous = lastNode
              Next = None }

    linkedList.FirstNode <-
        linkedList.FirstNode
        |> Option.orElse linkedList.LastNode

    lastNode
    |> Option.iter (fun node -> node.Next <- linkedList.LastNode)

let unshift newValue linkedList =
    let firstNode = linkedList.FirstNode

    linkedList.FirstNode <-
        Some
            { Value = newValue
              Next = firstNode
              Previous = None }

    linkedList.LastNode <-
        linkedList.LastNode
        |> Option.orElse linkedList.FirstNode

    firstNode
    |> Option.iter (fun node -> node.Previous <- linkedList.FirstNode)
